src/realm/object-store/c_api/realm.cpp

```diff
@@ -173,6 +179,23 @@ RLM_API realm_callback_token_t* realm_add_realm_changed_callback(realm_t* realm,
         realm, CBindingContext::get(*realm).realm_changed_callbacks().add(std::move(func)));
 }
 
+RLM_API realm_refresh_callback_token_t* realm_add_realm_refresh_callback(realm_t* realm,
+                                                                         realm_on_realm_refresh_func_t callback,
+                                                                         realm_userdata_t userdata,
+                                                                         realm_free_userdata_func_t userdata_free)
+{
+    util::UniqueFunction<void()> func = [callback, userdata = UserdataPtr{userdata, userdata_free}]() {
+        callback(userdata.get());
+    };
+    const util::Optional<DB::version_type>& latest_snapshot_version = (*realm)->latest_snapshot_version();
+
+    if (!latest_snapshot_version)
+        return nullptr;
+
```

Comment:
I think this might be a good place to check `if (realm->is_frozen()) ` and return an error or nullptr.